docs/xks/developer-guide/container-security.md

```diff
@@ -0,0 +1,293 @@
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  containers:
+    - name: nginx
+      image: nginx:1.14.2
+      securityContext:
+        readOnlyRootFilesystem: true
+        allowPrivilegeEscalation: false
+        capabilities:
+          drop:
+            - NET_RAW
+            - CAP_SYS
+```
+
+### Read Only Root Filesystem
+
+This setting is the one that may have the largest impact on a deployed application if it is expected to write to the file system. Containers may be ephemeral but that does not mean that certain applications may still want to write temporary files. Enabling read only root filesystem blocks the container from writing to its filesystem by default. The reasoning for this is that if an attacker is able to gain access to a container the attacker will not be able to modify any files or add any additional binaries to the container.
+
+Of course there is a solution for applications that still do want to write to the filesystem. The solution is to create a [`emptyDir`](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume and mount it to the directory to which files should be written. In the example below the `emptyDir` volume `tmp` is mounted to the path `/tmp` making it possible to write files to the directory.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  containers:
+    - name: nginx
+      image: nginx:1.14.2
+      securityContext:
+        readOnlyRootFilesystem: true
+        allowPrivilegeEscalation: false
+        capabilities:
+          drop:
+            - NET_RAW
+            - CAP_SYS_ADMIN
+      volumeMounts:
+        - mountPath: /tmp
+          name: tmp
+  volumes:
+    - emptyDir: {}
+      name: tmp
+```
+
+### Allow Privilege Escalation
+
+Disabling privilege escalation for all Pods is critical for applications to to gain greater access to the Node than they need to. This setting will probably never affect most applications.
+
+### Dropped Capabilities
+
+Linux capabilities are certain root permissions that can be given to containers without giving root access. Certain capabilities should never be given to a container as they would give too much access, for this reason they are always dropped.
+
+* `CAP_SYS_ADMIN` - Is the same as making the container root, would allow a container to escalate privileges.
+* `NET_RAW` - Allows a container to craft raw network packets which can be exploited for malicious actions.
+
+## Pod Constraints
+
+In addition to the default security context there are certain configuration settings which are not allowed in XKF. They have a lower impact on end users as they are opted into, but they are good to be aware of and can explain why a Pods creation may be blocked by the cluster. The enforcement is done before the Pod is created, this means that a Deployment may be allowed to be created but the Pods that result will not, causing a Deployment to always have zero replicas.
```

Comment:
"causing a Deployment to always have zero replicas." I know that we are not rewriting the k8s docs but should we find a good link that explain debugging deployments and have that a link to help the developers to understand why there deployment isn't starting a new pod.

Reply:
Could not find a good link, do you have one? Its difficult moving between rewriting docs and adding our own oppinion about things.